samples/s3-ls/s3-ls.c

```diff
@@ -0,0 +1,228 @@
+    app_ctx.c_var = (struct aws_condition_variable)AWS_CONDITION_VARIABLE_INIT;
+    aws_mutex_init(&app_ctx.mutex);
+
+    aws_s3_library_init(allocator);
+
+    if (enable_logs) {
+        setup_logger(&app_ctx);
+    }
+
+    // credentials
+    const char* access_key_id = getenv("AWS_ACCESS_KEY_ID");
+    struct aws_credentials_provider* credentials_provider = NULL;
+
+    if (access_key_id != NULL) {
+        // use credentials from environment variables
+        struct aws_credentials_provider_environment_options environmentOptions;
+        AWS_ZERO_STRUCT(environmentOptions);
+        credentials_provider = aws_credentials_provider_new_environment(allocator, &environmentOptions);
+    }
+
+    if (credentials_provider == NULL) {
+        printf("ERROR: no credentials found. This sample currently requires the credentials to be specified in environment variables.\n");
+        return -1;
+    }
+
+    // signing config
+    aws_s3_init_default_signing_config(&app_ctx.signing_config, aws_byte_cursor_from_c_str(region), credentials_provider);
+    app_ctx.signing_config.flags.use_double_uri_encode = false;
+
+    // event loop
+    struct aws_event_loop_group* event_loop_group = aws_event_loop_group_new_default(allocator, 0, NULL);
+
+    // resolver
+    struct aws_host_resolver_default_options resolver_options = {
+            .el_group = event_loop_group,
+            .max_entries = 8,
+    };
+    struct aws_host_resolver *resolver = aws_host_resolver_new_default(allocator, &resolver_options);
+
+    // client bootstrap
+    struct aws_client_bootstrap_options bootstrap_options = {
+            .event_loop_group = event_loop_group,
+            .host_resolver = resolver,
+    };
+    struct aws_client_bootstrap* client_bootstrap = aws_client_bootstrap_new(allocator, &bootstrap_options);
+    if (client_bootstrap == NULL) {
+        printf("ERROR initializing client bootstrap\n");
+        return -1;
+    }
+
+    // s3 client
+    struct aws_s3_client_config client_config;
+    AWS_ZERO_STRUCT(client_config);
+    client_config.client_bootstrap = client_bootstrap;
+    client_config.region = aws_byte_cursor_from_c_str(region);
+    client_config.signing_config = &app_ctx.signing_config;
+    struct aws_s3_client* client = aws_s3_client_new(allocator, &client_config);
+
+    // listObjects
+    struct aws_s3_list_bucket_v2_params params;
+    AWS_ZERO_STRUCT(params);
+    params.client = client;
+    params.bucket_name = bucket;
+    params.prefix = prefix;
+
+    char endpoint[1024];
+    snprintf(endpoint, sizeof(endpoint), "s3.%s.amazonaws.com", region);
+    params.endpoint = aws_byte_cursor_from_c_str(endpoint);
+    params.user_data = &app_ctx;
+    params.on_object = &s_on_object;
+    params.on_list_finished = &s_on_list_finished;
+
+    struct aws_s3_paginator* paginator = aws_s3_initiate_list_bucket(allocator, &params);
+    int paginator_result = aws_s3_paginator_continue(paginator, &app_ctx.signing_config);
```

Comment:
feel free to use the blocking function i wrote, it would save you all this predicate/ cvar dance by doing it for you